abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+
+#### `proxy_resume_stream`
+
+* params:
+  - `i32 (proxy_stream_type_t) stream_type`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resume processing of paused downstream or upstream (`stream_type`) connection
+in current context.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `stream_type`.
+
+Returns `InvalidOperation` when stream wasn't paused.
+
+#### `proxy_close_stream`
+
+* params:
+  - `i32 (proxy_stream_type_t) stream_type`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Close downstream or upstream (`stream_type`) connections in current context.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `stream_type`.
+
+Returns `InvalidOperation` when stream is already closed.
+
+
+## HTTP (L7) extensions
+
+Note: The same unique context identifier (`stream_id`) is shared by a pair of
+      HTTP request and response.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_http_request_headers`
+
+* params:
+  - `i32 (uint32_t) stream_id`
+  - `i32 (size_t) num_headers`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (proxy_action_t) next_action`
+
+Called when HTTP request headers for stream `stream_id` are received from
+downstream.
+
+Headers (`num_headers` entries) can be retrieved using `proxy_get_map_values`.
+
+Return value (`next_action`) instructs the host environment what action to take:
+- `Continue` means that this HTTP request should be accepted and that HTTP
+  headers should be forwarded upstream.
+- `EndStream` means that this HTTP request should be accepted, that HTTP headers
+  should be forwarded upstream and that the underlying stream should be closed
+  afterwards.
+- `Done` means that this HTTP request should be accepted, that HTTP headers
+  should be forwarded upstream and that this extension shouldn't be called again
+  for this HTTP request.
+- `Pause` instructs the host environment to pause processing until
+  `proxy_resume` is called.
+- `Close` means that the stream should be closed immediately.
+
+#### `proxy_on_http_request_body`
+
+* params:
+  - `i32 (uint32_t) stream_id`
+  - `i32 (size_t) body_size`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (proxy_action_t) next_action`
+
+Called for each chunk of HTTP request body for stream `stream_id` received from
+downstream.
+
+Request body (of size `body_size`) can be retrieved using `proxy_get_buffer`.
+
+Return value (`next_action`) instructs the host environment what action to take:
+- `Continue` means that this HTTP request body chunk should be forwarded
+  upstream.
+- `EndStream` means that this HTTP request body chunk should be forwarded
+  upstream and that the underlying stream should be closed afterwards.
+- `Done` means that this HTTP request body should be forwarded upstream and that
+  this extension shouldn't be called again for this HTTP request.
+- `Pause` means that further processing should be paused until `proxy_resume` is
```

Comment:
WaitForMoreData would call this callback when more data arrives where Pause would swallow that event until the wasm filter resumed.

I assume this maps to StopAllIterationAndBuffer, rather than StopIteration (which is kind of the Envoy default) or StopAllItreationAndWatermark.  Might be worth clarifying?

Reply:
In general it seems like we are missing the options around watermarking vs. buffering. I think that is needed. I commented on this elsewhere.